### ISSUE TITLE
ParametersHolder.get() works only for exact types (causes DefinitionParameterException)

### ISSUE THREAD

**Issue by j2esu** [InsertKoinIO/koin#1121]
**Describe the bug**

I've encountered a crashes when migrating from destructured params. It seems like ParametersHolder.get() works only for exact types. For example, I've got

> org.koin.core.error.DefinitionParameterException: No value found for type 'java.util.List'

when passing any list implementation where List required. It works ok when using get(index), but I think it's less convenient

**To Reproduce**
```
module {
      factory {
             Foo(it.get())
      }
}

class Foo(list: List<String>)

get<Foo> { parametersOf(listOf("")) } // No value found for type 'java.util.List'

```

**Koin project used and used version (please complete the following information):**

io.insert-koin:koin-android:3.1.0

**Comment by almozavr** [comment]
I'd say this is super critical, as there is no way to get parameter abstraction

**Comment by almozavr** [comment]
Probably this fix would be a change for `ParametersHolder`

```
fun <T> getOrNull(clazz: KClass<*>): T? {
    return listOf<Any?>().firstNotNullOfOrNull { value ->
        if (value != null && value::class.isSubclassOf(clazz)) value as? T? else null
    }
}
```

**Comment by felipepedroso** [comment]
I'm also reproducing this issue on `2.2.3` 😢

**Comment by arnaudgiuliani** [comment]
Can you check in 3.1.1?

**Comment by almozavr** [comment]
@arnaudgiuliani yeap, it works now, but I have to notice, it's confusing that `ParametersHolder`'s `get` uses concrete `getOrNull(clazz: KClass<*>)` and not `<reified T> getOrNull()` -> because `elementAt` works the same as reified variant, so it create inconsistent behaviour:
- works: `factory<SplashNavigator> { (activity: Activity) -> SplashScreenAppNavigator(activity) }`
- not works: `factory<SplashNavigator> { SplashScreenAppNavigator(get()) }` // tries to get concrete type by default

Where 
```
class SplashScreenAppNavigator(
    private val activity: Activity // not a concrete type
) : SplashNavigator {
```

```
class SplashScreenActivity : AppCompatActivity() {

    private val navigator: SplashNavigator by inject { parametersOf(this) }
```

**Comment by almozavr** [comment]
I'd propose to rename `getOrNull(clazz: KClass<*>)` to `getExactlyOrNull(clazz: KClass<*>)`  and also change `get()` to 
```
inline fun <reified T : Any> get(): T = getOrNull() ?: throw DefinitionParameterException("No value found for type '${T::class.getFullName()}'")
```

**Comment by j2esu** [comment]
> Can you check in 3.1.1?

I've checked an example in the description, still not working

**Comment by tmoura10** [comment]
I was about to create the same issue. The current workaround in `3.1.1` is to use `getOrNull()!!`.

I agree with @almozavr that maybe we can only `get()` to use the definition that was added [here](https://github.com/InsertKoinIO/koin/commit/90d8ec31b5a29653208100b4eed1fc30c126aa88) plus change the `getOrNull(clazz: KClass<*>)` to have a check for `isInstance`

```kotlin
open fun <T> getOrNull(clazz: KClass<*>): T? {
    return _values.firstNotNullOfOrNull { value -> if (clazz.isInstance(value)) value as T? else null }
}

inline fun <reified T : Any> get(): T = getOrNull() ?: throw IllegalArgumentException("No value found for type '${T::class}'")
```

**Comment by arnaudgiuliani** [comment]
Yep, I'll add a fix in 3.1.2 with following:

```kotlin
open fun <T> getOrNull(clazz: KClass<*>): T? {
        return _values.firstNotNullOfOrNull { value -> if (clazz.isInstance(value)) value as? T? else null }
    }
```

Thanks 👍
